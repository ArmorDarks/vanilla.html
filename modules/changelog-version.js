const { exec } = require('child_process')
const { readFile, writeFileSync } = require('fs')
const { join } = require('path')
const moment = require('moment')
const { version, repository } = require(join(process.cwd(), 'package.json'))
const { green, red } = require('chalk')

/**
 * Update `CHANGELOG.md` HEAD to the latest version,
 * with proper compare URL and date, as well as
 * add new `HEAD` header with proper URL.
 *
 * CHANGELOG should follow [keepchangelog.com specs](https://keepachangelog.com).
 *
 * Relies on `package.json` `version` and `repository.url` values.
 *
 * Sould be run in `version` NPM script.
 */

const logError = (message) => console.error(red(message))

if (!repository.url || !repository.url.includes('github.com') || !repository.url.endsWith('.git')) {
  logError('[changelog-version] please, specify `package.json` `repository.url` property with a valid Github URL')
  process.exit(1)
}

const repositoryURL = repository.url.slice(0, -4)
const date = moment().format('YYYY-MM-DD')

const CHANGELOG_FILENAME = 'CHANGELOG.md'
const HEAD_PATTERN = /^## \[?(HEAD|Unreleased)\]?$/m
const VERSION_PATTERN = /^## \[?(\d+\.\d+\.\d+)/m

readFile(CHANGELOG_FILENAME, { encoding: 'utf-8' }, (error, changelog) => {
  if (error) return console.log(`[changelog-version] seems to be no ${CHANGELOG_FILENAME}, skipping changelog version update`)

  const hasHead = HEAD_PATTERN.test(changelog)

  if (!hasHead) {
    logError('[changelog-version] changelog does not have HEAD version, can not update it with latest version')
    process.exit(1)
  }

  const previousVersion = changelog.match(VERSION_PATTERN)[1]
  const releaseHeader = `## [${version}](${repositoryURL}/compare/v${previousVersion}...v${version}) - ${date}`
  const newHEADHeader = `## [HEAD](${repositoryURL}/compare/v${version}...HEAD)`
  const updatedChangelog = changelog
    .replace(HEAD_PATTERN, `${newHEADHeader}\n\n## HEAD`)
    .replace(HEAD_PATTERN, releaseHeader)

  writeFileSync(CHANGELOG_FILENAME, updatedChangelog)

  exec(`git add ${CHANGELOG_FILENAME}`, (error, stdout) => {
    if (error) {
      logError('[changelog-version] An error occurred while staging updated changelog:\n')
      logError(error)
      process.exit(1)
    }

    console.log(green(`\n[changelog-version] succesfully updated changelog the the v${version}`))
  })
})
