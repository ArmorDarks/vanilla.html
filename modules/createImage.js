const sharp = require('sharp')
const { mkdirSync, existsSync } = require('fs')
const { resolve, join, parse } = require('path')

const mock = (path) => {
  const image = sharp(path)
  image.done = async () => {
    return Promise.resolve(await image.metadata())
  }
  return image
}

const createImage = (
  src,
  {
    prefix = '',
    filename,
    postfix = '',
    outputDir,
    cwd = '',
    build,
    override
  }
) => {
  if (typeof src !== 'string') throw new Error(`[createImage] \`src\` should be a string, \`${typeof src}\` provided`)
  if (!build) return mock()

  let { dir, name, ext } = parse(src)

  if (outputDir) dir = outputDir

  const absDir = resolve(cwd, dir)
  const absSrc = resolve(cwd, src)

  if (!existsSync(absSrc)) throw new Error(`[createImage] src \`${absSrc}\` does not exist`)

  if (filename) name = filename

  const file = `${prefix}${name}${postfix}${ext === '.svg' ? '.png' : ext}`
  const output = join(dir, file)
  const absOutput = resolve(output)

  if (!override && existsSync(absOutput)) return mock(absOutput)
  if (!existsSync(absDir)) mkdirSync(absDir)

  const image = sharp(absSrc, { failOnError: true })

  const data = {
    src: output,
    prefix,
    filename,
    postfix,
    outputDir,
    cwd,
    build
  }

  image.done = async () => {
    let metadata

    try {
      metadata = await image.toFile(absOutput)
    } catch (error) {
      error.message = `[createImage]: ${error.message}`
      throw error
    }

    return Promise.resolve(Object.assign(data, metadata))
  }

  return image
}

module.exports = createImage
