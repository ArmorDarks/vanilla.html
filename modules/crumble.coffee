_           = require('lodash')
{ extname } = require('path')

###*
 * Explodes string path into array breadcrumb
 * @param  {string} to Relative or absolute path
 * @return {array}     Array, which consists of path's fragments
 * @example crumble('/url/to/index.html') -> `['url', 'to']`
 * @example crumble('/url/to/404.html') -> `['url', 'to', '404']`
 * @example crumble('/') -> `['index']`
###
module.exports = (to) ->
  if to == '/'
    breadcrumb = ['index']
  else
    breadcrumb = _.chain(to).trimStart('/').trimEnd('/').trimEnd(extname(to)).split('/').value()

    if breadcrumb.length >= 2 and _.last(breadcrumb) == 'index' then breadcrumb.pop()

  return breadcrumb