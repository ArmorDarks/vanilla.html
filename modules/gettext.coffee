_           = require('lodash')
path        = require('path')
NodeGettext = require('node-gettext')
gettext     = new NodeGettext()

module.exports = (grunt) ->
  return class Gettext

    ###*
     * Load l10n files and make l10n class available to Grunt-related tasks
     * @note Do not use 'LC_MESSAGES' in path to locales
     * @todo Since node-gettext doesn't have method for switching between languages AND domains,
     *       use `dgettext('{{locale}}:{{domain'), 'String')` to switch between locales and domains
     *       `/locale/en/{defaultLocale}.po` will result in `en` domain.
     *       `/locale/en/nav/bar.po` will result in `en:nav:bar` domain.
     *       Related Github issues:
     *       * https://github.com/andris9/node-gettext/issues/22
     *       * https://github.com/LotusTM/Kotsu/issues/45
    ###
    constructor: (@locales, @localesDir) ->
      @locales.forEach (locale) =>

        grunt.file.expand({ cwd: @localesDir + '/' + locale, filter: 'isFile' }, '**/*.po').forEach (filepath) =>
          defaultDomain = 'messages'

          domain   = filepath.replace('LC_MESSAGES/', '').replace('/', ':').replace(path.extname(filepath), '')
          domain   = if domain == defaultDomain then locale else locale + ':' + domain
          messages = grunt.file.read(@localesDir + '/' + locale + '/' + filepath, { encoding: null })

          gettext.addTextdomain(domain, messages)

    installNunjucksGlobals: (env, currentLocale) ->

      # --------------
      # i18n functions
      # --------------

      ###*
      * Load string from current locale
      * @param {string} string          String, which should be loaded
      * @return {string} Translated string into current locale
      ###
      env.addGlobal 'gettext', (string) ->
        gettext.dgettext(currentLocale, string)

      ###*
       * Load string from specified domain
       * @param {string} domain = currentLocale Domain or locale, from which string should be loaded
       *                                        `en-US:other:inner` will load from `en-US/other/inner.po`
       *                                        `:other` will load from `{{currentLocale}}/other.po`
       * @param {string} string                 String, which should be loaded
       * @return {string} Translated string into specified locale
      ###
      env.addGlobal 'dgettext', (domain = currentLocale, string) ->
        domain = if domain.charAt(0) == ':' then currentLocale + domain else domain
        gettext.dgettext(domain, string)

      ###*
      * Load plural string from current locale
      * @param {string} string          String, which should be loaded
      * @param {string} pluralString    Plural form of string
      * @param {number} count           Count for detecting correct plural form
      * @return {string} Pluralized and translated into current locale string
      ###
      env.addGlobal 'ngettext', (string, pluralString, count) ->
        gettext.dngettext(currentLocale, string, pluralString, count)

      ###*
       * Load plural string from specified domain
       * @param {string} domain = currentLocale Domain or locale, from which string should be loaded
       *                                        `en-US:other:inner` will load from `en-US/other/inner.po`
       *                                        `:other` will load from `{{currentLocale}}/other.po`
       * @param {string} string                 String, which should be loaded
       * @param {string} pluralString           Plural form of string
       * @param {number} count                  Count for detecting correct plural form
       * @return {string} Pluralized and translated into specified loca stringle
      ###
      env.addGlobal 'dngettext', (domain = currentLocale, string, pluralString, count) ->
        domain = if domain.charAt(0) == ':' then currentLocale + domain else domain
        gettext.dngettext(domain, string, pluralString, count)

      ###*
      * Load string of specific context from current locale
      * @param {string} context         Context of curret string
      * @param {string} string          String, which should be loaded
      * @return {string} Translated string into current locale
      ###
      env.addGlobal 'pgettext', (context, string) ->
        gettext.dpgettext(currentLocale, context, string)

      ###*
       * Load string of specific context from specified domain
       * @param {string} domain = currentLocale Domain or locale, from which string should be loaded
       *                                        `en-US:other:inner` will load from `en-US/other/inner.po`
       *                                        `:other` will load from `{{currentLocale}}/other.po`
       * @param {string} context                Context of curret string
       * @param {string} string                 String, which should be loaded
       * @return {string} Translated string into specified locale
      ###
      env.addGlobal 'dpgettext', (domain = currentLocale, context, string) ->
        domain = if domain.charAt(0) == ':' then currentLocale + domain else domain
        gettext.dpgettext(domain, context, string)

      ###*
      * Load plural string of specific context from current locale
      * @param {string} context         Context of curret string
      * @param {string} string          String, which should be loaded
      * @param {string} pluralString    Plural form of string
      * @param {number} count           Count for detecting correct plural form
      * @return {string} Pluralized and translated into current locale string
      ###
      env.addGlobal 'npgettext', (context, string, pluralString, count) ->
        gettext.dnpgettext(currentLocale, context, string, pluralString, count)

      ###*
       * Load plural string of specific context from specified domain
       * @param {string} domain = currentLocale Domain or locale, from which string should be loaded
       *                                        `en-US:other:inner` will load from `en-US/other/inner.po`
       *                                        `:other` will load from `{{currentLocale}}/other.po`
       * @param {string} context                Context of curret string
       * @param {string} string                 String, which should be loaded
       * @param {string} pluralString           Plural form of string
       * @param {number} count                  Count for detecting correct plural form
       * @return {string} Pluralized and translated into specified loca stringle
      ###
      env.addGlobal 'dnpgettext', (domain = currentLocale, context, string, pluralString, count) ->
        domain = if domain.charAt(0) == ':' then currentLocale + domain else domain
        gettext.dnpgettext(domain, context, string, pluralString, count)

      return