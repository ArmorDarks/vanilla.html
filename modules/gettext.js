const { join, extname } = require('path')
const NodeGettext = require('node-gettext')
const { file: { expand, read } } = require('grunt')

module.exports = class Gettext {
  constructor (opts) {
    if (!opts.cwd) {
      throw new Error('[gettext] argument `options.cwd` should be privided')
    }

    const options = Object.assign({
      defaultDomain: 'messages'
    }, opts)

    this.cwd = options.cwd
    this.defaultDomain = options.defaultDomain
    this.localesDirs = expand({ cwd: this.cwd, filter: 'isDirectory' }, '*', '!*templates')

    /**
     * Selects corresponding to specified locale Gettext instance, otherwise creates new.
     * @param {string} locale Name of locale which should be invoked or created
    */
    this.setLocale = function (locale) {
      if (!this[locale]) {
        this[locale] = new NodeGettext()
        this[locale].locale = locale
      }

      this.gt = this[locale]
    }

    /**
     * Binds new domain to current active locale.
     * Expects your l10n files to be under `{localeName}/LC_MESSAGES/..` or `{localeName}/..` paths.
     * @param  {string} domain       Package name for new domain
     * @param  {string} [cwd] = @cwd Full path to locales dir
    */
    this.bindTextdomain = function (domain, cwd = this.cwd) {
      cwd = join(cwd, this.gt.locale)
      // Discover possible file names and select first matching
      const domainFile = expand({ cwd, filter: 'isFile' }, `{,LC_MESSAGES/}${domain}.{mo,po}`)[0]
      const domainPath = join(cwd, domainFile)
      const messages = read(domainPath, {encoding: null})

      this.gt.addTextdomain(domain, messages)
    }

    /**
     * Load l10n files and bind them to domains based on filepaths for current locale.
     * @param  {string} [cwd] = @cwd Full path to locales dir
     * @example `/locales/en/foo/bar.po` will result in `foo/bar` domain.
    */
    this.autobindTextdomains = function (cwd = this.cwd) {
      cwd = join(cwd, this.gt.locale)

      expand({ cwd, filter: 'isFile' }, '{,**/}*.{mo,po}').forEach(domainpath => {
        const domain = domainpath.replace(extname(domainpath), '')
        return this.bindTextdomain(domain)
      })
    }

    // Load existing l10n files as locales
    this.localesDirs.forEach((locale) => {
      this.setLocale(locale)
      this.autobindTextdomains()
    })

    // Ensure that no locale set as active during init
    this.gt = null

    this.setTextdomain = (domain = this.defaultDomain) => this.gt.textdomain(domain)

    // See https://github.com/alexanderwallin/node-gettext/tree/master#translation-methods
    this.gettext = (message) => this.gt.gettext(...arguments)
    this.dgettext = (domain, message) => this.gt.dgettext(...arguments)
    this.ngettext = (message, pluralMessage, count) => this.gt.ngettext(...arguments)
    this.dngettext = (domain, message, pluralMessage, count) => this.gt.dngettext(...arguments)
    this.pgettext = (context, message) => this.gt.pgettext(...arguments)
    this.dpgettext = (domain, context, message) => this.gt.dpgettext(...arguments)
    this.npgettext = (context, message, pluralMessage, count) => this.gt.npgettext(...arguments)
    this.dnpgettext = (domain, context, message, pluralMessage, count) => this.gt.dnpgettext(...arguments)

    this.nunjucksExtensions = (env, currentLocale) => {
      this.setLocale(currentLocale)
      this.setTextdomain(this.defaultDomain)

      env.addGlobal('setLocale', (locale = currentLocale) => this.setLocale(locale))
      env.addGlobal('setTextdomain', (domain = this.defaultDomain) => this.textdomain(...arguments))

      env.addGlobal('gettext', (message) => this.gettext(...arguments))
      env.addGlobal('dgettext', (domain, message) => this.dgettext(...arguments))
      env.addGlobal('ngettext', (message, pluralMessage, count) => this.ngettext(...arguments))
      env.addGlobal('dngettext', (domain, message, pluralMessage, count) => this.dngettext(...arguments))
      env.addGlobal('pgettext', (context, message) => this.pgettext(...arguments))
      env.addGlobal('dpgettext', (domain, context, message) => this.dpgettext(...arguments))
      env.addGlobal('npgettext', (context, message, pluralMessage, count) => this.npgettext(...arguments))
      env.addGlobal('dnpgettext', (domain, context, message, pluralString, count) => this.dnpgettext(...arguments))
    }
  }
}
