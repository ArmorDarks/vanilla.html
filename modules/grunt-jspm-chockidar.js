const spawn = require('cross-spawn')
const { resolve } = require('path')
const { red, cyan } = require('chalk')

let firstLaunch = true

module.exports = ({ registerMultiTask, log }) =>
  registerMultiTask('jspmChockidar', 'Launch chockidar-socket-emitter for JSPM (SystemJS)', function () {
    const done = this.async()
    const { cwd } = this.options({
      cwd: ''
    })

    log.ok('Launching chockidar-socket-emitter...')

    const chockidar = spawn(
      resolve('node_modules/.bin/chokidar-socket-emitter'),
      {
        cwd,
        env: { FORCE_COLOR: true }
      }
    )

    chockidar.stdout.pipe(process.stdout)

    chockidar.stdout.on('data', (data) => {
      if (data.toString().includes('chokidar-socket-emitter listening on')) done()
    })

    chockidar.stderr.on('data', (error) => {
      console.log(red(error.toString()))

      if (firstLaunch) done(new Error('chockidar-socket-emitter failed.'))

      firstLaunch = false

      done()
    })
  })
