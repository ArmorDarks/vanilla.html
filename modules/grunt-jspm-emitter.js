const chokidarEventEmitter = require('chokidar-socket-emitter')

module.exports = ({ registerMultiTask, log }) =>
  registerMultiTask('jspmEmitter', 'Launch chockidar-socket-emitter for JSPM (SystemJS)', function () {
    const done = this.async()
    const { cwd, watchTask } = this.options({
      cwd: '.'
    })

    log.ok('Launching chockidar-socket-emitter...')

    chokidarEventEmitter({ path: cwd, relativeTo: cwd }, () => {
      if (watchTask) done()
    })
  })
