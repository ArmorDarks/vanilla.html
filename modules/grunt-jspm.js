const { spawn } = require('child_process')
const { red, cyan, yellow } = require('chalk')

/**
 * This task exists only because as of now JSPM does not expose its watch API
 * In other words, watch mode (`-wid`) flag can be launched only through CLI
 * This task workarounds this issue and allows to run JSPM watch in parallel
 * with other Grunt tasks.
 *
 * Besides, it allows to watch any amount of builds in parallel.
 *
 * Task target parameters:
 *
 * @param {object[]} files                 List of files to build or watch
 * @param {string}   files[].dest          Build file destination
 * @param {string}   [files[].packageName] SystemJS package name to build. Should
 *                                         be specified unless `files.src` used
 * @param {string}   [files[].src]         JS file to build. Should be specified
 *                                         unless `files.packageName` used
 * @param {object}   [options]             Task options
 * @param {string[]} [options.args]        List of args to call.
 *                                         For instance, `-wid` or `--minify`
 */

let firstLaunch = true

module.exports = ({ registerMultiTask, log, util: { pluralize } }) =>
  registerMultiTask('jspm', 'Launch JSPM', function () {
    const done = this.async()

    const end = (error) => {
      log.error(error)
      done()
    }

    const { args } = this.options({
      args: []
    })

    if (!this.files.length) return end('No files specified.')

    log.ok(`Building ${cyan(this.files.length)} ${pluralize(this.files.length, 'file/files')}...`)

    this.files.forEach(({ src, dest, packageName }) => {
      if (!dest) return end('No dest file specified')

      const source = packageName || src[0]

      if ((src && !src.length) && !packageName) return end('No src file or `packageName` specified')
      if (src && src.length > 1) return end('Only 1 src supported')

      process.env['FORCE_COLOR'] = true

      let buildArgs = [
        'run',
        'jspm',
        '--',
        'build',
        source,
        dest
      ]

      if (args && args.length) {
        buildArgs = buildArgs.concat(args)
      }

      const jspm = spawn(
        'npm',
        buildArgs,
        { shell: true }
      )

      jspm.stdout.pipe(process.stdout)

      jspm.stdout.on('data', (data) => {
        if (data.toString().includes('Built into')) done()
      })

      jspm.stderr.on('data', (error) => {
        const errorString = error.toString()
        const isWarning = /^(\s+)?(warn|npm warn|npm$|watchman|(.+ExperimentalWarning))/i.test(errorString)

        console.log(isWarning ? yellow(errorString) : red(errorString))

        if (isWarning) return
        if (firstLaunch) done(new Error('JSPM failed.'))

        firstLaunch = false

        done()
      })
    })
  })
