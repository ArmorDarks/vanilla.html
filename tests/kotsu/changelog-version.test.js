/* eslint-env jest */

import { updateVersion } from '../../modules/changelog-version'

const CHANGELOG_MOCK = `
# Changelog

## [HEAD](https://github.com/LotusTM/Kotsu/compare/v1.0.0...HEAD)

### Changed
- Something changed

## [1.0.0](https://github.com/LotusTM/Kotsu/compare/v0.1.0...v1.0.0) - 2018-03-30

### Added
- Something added
`

describe('`updateVersion` function', () => {
  it('should update version', () => {
    expect(updateVersion({ input: CHANGELOG_MOCK, version: '1.1.0', date: '2018-04-01' })).toMatchSnapshot()
  })
})
