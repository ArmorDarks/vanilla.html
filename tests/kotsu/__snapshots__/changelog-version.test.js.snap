// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`updateVersion\` function should update version 1`] = `
"
# Changelog

## [HEAD](https://github.com/LotusTM/Kotsu/compare/v1.1.0...HEAD)

## [1.1.0](https://github.com/LotusTM/Kotsu/compare/v1.0.0...v1.1.0) - 2018-04-01

### Changed
- Something changed

## [1.0.0](https://github.com/LotusTM/Kotsu/compare/v0.1.0...v1.0.0) - 2018-03-30

### Added
- Something added
"
`;
